### PR TITLE
haskellPackages.mpi-hs*: fix mpich vs openmpi linking

### PR DESCRIPTION
**Pull request by sheepforce**
## Description of changes

The `haskellPackages.mpi-hs*` packages by default link against `-lmpich` but in nixpkgs the default is OpenMPI in the `mpi` attribute. Due to the mechanism how `cabal2nix` passes dependencies in, the library is linked against MPICH libraries but gets an OpenMPI `mpiexec`, so that all Haskell-MPI programmes complain about unitialised environment when run (only OpenMPI runtime is initialised but MPICH runtime is used incorrectly by the Haskell package). This PR fixes the wrong linking and ensures correct behaviour of the test phase.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/haskell-modules/configuration-nix.nix
Comment (cdepillabout) [inline pkgs/development/haskell-modules/configuration-nix.nix:1344]: It seems like `mpi` has pkgconfig files available.

Maybe you could send a PR upstream making the .cabal files use the pkgconfig files?  That would be easier than all these overrides.  It is definitely a code smell that upstream hard-codes links to `/usr` in their `.cabal` file.

--------------------------------------------

It is strange having to override all of `librarySystemDepends`, `executableSystemDepends`, `testSystemDepends`, `extraLibraries`, and `configureFlags`.  I would have expected you to only need to override something like the `buildDepends` argument.

Also, you'll probably want to remove the `mpich` argument from the generated derivation (and possibly replace it with `mpi`?), which you can do in a PR to `cabal2nix`.

**Review by cdepillabout — CHANGES_REQUESTED** [review]

File: pkgs/development/haskell-modules/configuration-nix.nix
Comment (cdepillabout) [reply pkgs/development/haskell-modules/configuration-nix.nix:1344]: Oh, and one other thing: we generally ask for Haskell-related PRs to be made to the `haskell-updates` branch.  Could you rebase this PR on `haskell-updates`?

Comment (sheepforce) [reply pkgs/development/haskell-modules/configuration-nix.nix:1344]: > Maybe you could send a PR upstream making the .cabal files use the pkgconfig files? 

I can try tomorrow and will link. But the upstream project seems a little inactive, last commit is from beginning 2020, so let's see :)

`configureFlags` in the current setup is required to get the `-fopenmpi-debian` flag, which actually allow to link generically against `-lmpi` working for all major MPI implementations.

Wrt to the other overrides it is very well possible that I'm not understanding the haskell setup in nixpkgs well enough and maybe you can give me some hints there and comment on them? :slightly_smiling_face: 

- I've added mpi it to `executableSystemDepends` as the `mpiexec` executable will be required at runtime for every programme and library using `mpi-hs` packages
- I think `mpi` is not passed properly to the generated derivation. As far as I understand, `cabal2nix` is called first and determines which packages are passed to the nix expression building the actual haskell package, correct? I think at this stage `mpich` is still passed, because this is the default flag in the cabal file. Thus, when I override the `configureFlags` later to use `-fopenmpi-debian`, which switches to `mpi` instead of the `mpihs` dependency, `mpich` is already passed because it happened a stage earlier. Do I understand this correctly? :grimacing: 
- Is it enough to just pass `pkgs.mpi` to `buildDepends`? During building `-lmpi` is required, for compilation probably the `mpicc` compiler wrapper and during the test phase **and** runtime `mpiexec`.

Comment (sheepforce) [reply pkgs/development/haskell-modules/configuration-nix.nix:1344]: > Also, you'll probably want to remove the mpich argument from the generated derivation (and possibly replace it with mpi?), which you can do in a PR to cabal2nix. (Feel free to ask if you need pointers for this)

I need a pointer for this :smile: I assume it means this influences which attributes are passed to `mkDerivation` and I can get rid of `mpich` in first place?

Comment (sheepforce) [reply pkgs/development/haskell-modules/configuration-nix.nix:1344]: > Maybe you could send a PR upstream making the .cabal files use the pkgconfig files? That would be easier than all these overrides. It is definitely a code smell that upstream hard-codes links to /usr in their .cabal file.

https://github.com/eschnett/mpi-hs/pull/3 :slightly_smiling_face:

Comment (sheepforce) [reply pkgs/development/haskell-modules/configuration-nix.nix:1344]: While writing the PR upstream I've tried to nixify the mpi-hs repo directly and the best solution I've could come up with so far was something like this:

```nix
{ mkDerivation
, pkg-config
, base
, bytestring
, c2hs
, lib
, monad-loops
, mpi
, mpiCheckPhaseHook
, openssh
}:
mkDerivation {
  pname = "mpi-hs";
  version = "0.7.2.0";
  src = lib.cleanSource ./..;
  isLibrary = true;
  isExecutable = true;
  buildTools = [ pkg-config ];
  pkg-configDepends = [ mpi ];
  libraryHaskellDepends = [ base bytestring monad-loops ];
  libraryToolDepends = [ c2hs ];
  executableHaskellDepends = [ base ];
  executableSystemDepends = [ mpi ];
  testHaskellDepends = [ base monad-loops ];
  testSystemDepends = [ openssh mpi mpiCheckPhaseHook] ;
  homepage = "https://github.com/eschnett/mpi-hs#readme";
  description = "MPI bindings for Haskell";
  license = lib.licenses.asl20;
}
```
(manually edited result of `cabal2nix`. But currently I've no idea how to do something like this in the nixpkgs infrastructure :grimacing:

Comment (sheepforce) [reply pkgs/development/haskell-modules/configuration-nix.nix:1344]: > Oh, and one other thing: we generally ask for Haskell-related PRs to be made to the haskell-updates branch. Could you rebase this PR on haskell-updates?

Done :+1:

Comment (sheepforce) [reply pkgs/development/haskell-modules/configuration-nix.nix:1356]: With the upstream PR merged, do we need a new release on hackage or do we pull in the commits as patches?

Comment (cdepillabout) [reply pkgs/development/haskell-modules/configuration-nix.nix:1356]: The easiest thing for us would be to get a new release on Hackage.

Maybe you could ask to be a co-maintainer of the package?  That way you wouldn't have to ask the current maintainer for doing things like bumping versions, releasing to hackage, etc.

If the maintainer doesn't want to do a release to hackage, then the alternative is to apply patches in Nixpkgs (but we try to avoid this when possible).

Comment (sheepforce) [reply pkgs/development/haskell-modules/configuration-nix.nix:1356]: Alright, I'm second maintainer now and we have a new version on Hackage https://hackage.haskell.org/package/mpi-hs-0.7.3.0

Do we need to wait for the next Hackage update?

Comment (cdepillabout) [reply pkgs/development/haskell-modules/configuration-nix.nix:1356]: Nice, glad you got a new release out!

> Do we need to wait for the next Hackage update?

Yeah, the easiest thing to do is to wait for the next Hackage update, and then send a PR marking `mpi-hs` as unbroken (and possibly adding any overrides that are still necessary).

It should be the PR after https://github.com/NixOS/nixpkgs/pull/255037.

Comment (sheepforce) [reply pkgs/development/haskell-modules/configuration-nix.nix:1356]: Alright, so mpi-hs 0.7.3.0 has landed in #259060, although as `mpi-hs_0_7_3_0`. I was thinking `mpi-hs` should be somewhat like 

```nix
let addMpiHsTestDeps = drv: addTestToolDepends
      [ pkgs.openssh pkgs.mpiCheckPhaseHook ]
      drv;
in mpi-hs = addMpiHsTestDeps (
          if pkgs.mpi.pname == "openmpi"
          then callCabal2nixWithOptions "mpi-hs" mpi-hs_src "-fopenmpi -f-mpich -f-mvapich" { ompi = pkgs.mpi; }
          else if pkgs.mpi.pname == "mpich"
          then callCabal2nixWithOptions "mpi-hs" mpi-hs_src "-f-openmpi -fmpich -f-mvapich" { mpich = pkgs.mpi; }
          else if pkgs.mpi.pname == "mvapich2"
          then callCabal2nixWithOptions "mpi-hs" mpi-hs_src "-f-openmpi -f-mpich -fmvapich" { mvapich2 = pkgs.mpi; }
          else callCabal2nixWithOptions "mpi-hs" mpi-hs_src "-f-openmpi -f-mpich -f-mvapich" { mpi = pkgs.mpi; }
        );
```
However, I think I don't really understand how to call a package in this infrastructure with options. I couldn't find another place where `callCabal2nixWithOptions` is used in the haskell infrastructure within nixpkgs. Also what to do with mpi-hs vs mpi-hs_0_7_3_0? :)

Comment (maralorn) [reply pkgs/development/haskell-modules/configuration-nix.nix:1356]: callCabal2nix(withOptions) uses "import from derivation" and is therefore not allowed in nixpkgs.
We generally modify packages with the helpers listed in the nixpkgs manual first and foremost `overrideCabal`.

Comment (sheepforce) [reply pkgs/development/haskell-modules/configuration-nix.nix]: Alright, understandable. I came up with a different solution (and pushed it), but have to admit it feels somewhat hacky. As the expression is generated from the default flags, formally `mpi-hs` always has `ompi` (OpenMPI) as pkg-config dependency. However, the purpose here is exactly to make the tests pass and make it work with the other MPI implementations. Thus, I'm now changing cabal flags via `enableCabalFlag` and `disableCabalFlag` and override `ompi` to always use `pkgs.mpi`. However, the pkg-config dependency should of course correctly be `mpich` or `mvapich` in the other cases. But this seems a formality. Please have another look and let me know what you think :)

File: pkgs/development/haskell-modules/configuration-nix.nix
Comment (maralorn) [inline pkgs/development/haskell-modules/configuration-nix.nix:1357]: I am wondering whether this code is a bit to general in the mpiImpl. If the lib.optional case fails the package will be basically unusable, won’t it? Doesn’t that mean it would better to fail loudly in that case?

**Review by maralorn — APPROVED** [review]
I have one comment.

I think we can merge this when we have discussed that point.

File: pkgs/development/haskell-modules/configuration-nix.nix
Comment (sheepforce) [reply pkgs/development/haskell-modules/configuration-nix.nix:1357]: Luckily not! :) If none of the flags is true, mpi-hs will try to generically link against libmpi, see here https://github.com/eschnett/mpi-hs/blob/master/package.yaml#L89
This actually works for all three MPI implementations, too. If even this **would** fail (which it luckily doesn't) the test suite would not pass.

**Comment by maralorn** [comment]
Merged in:  9ebbc5f76505b30ddcf8a9dc75871be7404a4c4e

**Pull request closed without merging**
